File: leetcode/0020-valid-parentheses/solution.java
class Solution {
    public boolean isValid(String s) {
        Stack<Character> stack = new Stack<>();

        for(char c : s.toCharArray()){
            if(c == '{' || c == '[' || c == '('){
                stack.push(c);
            }
            else{
                if(stack.isEmpty()){
                    return false;
                }
                Character topOfStack = stack.peek();
                if (topOfStack == '{' && c == '}' || topOfStack == '(' && c == ')' || topOfStack == '[' && c == ']') {
                    stack.pop();
                } else {
                    return false;
                }
            }
        }

        return stack.isEmpty();
    }
}

File: leetcode/0020-valid-parentheses/solution.py
class Solution:
    def isValid(self, s: str) -> bool:
        stack = []
        for ch in s:
            if ch in "({[":
                stack.append(ch)
            else:
                if stack:
                    if stack[-1] == "(" and ch == ")":
                        stack.pop()
                    elif stack[-1] == "{" and ch == "}":
                        stack.pop()
                    elif stack[-1] == "[" and ch == "]":
                        stack.pop()
                    else:
                        return False
                else:
                    return False
        return not stack
